src/variable_rate.jl

```diff
@@ -0,0 +1,107 @@
+function total_variable_rate(jumps::JumpSet, u, p, t)
+    sum_rate = 0.0
+
+    vjumps = jumps.variable_jumps
+    if !isempty(vjumps)
+        for jump in vjumps
+            sum_rate += jump.rate(u, p, t)
+        end
+    end
+
+    return sum_rate
+end
+
+
+mutable struct GillespieIntegCallbackEventCache
+    prev_time::Float64
+    prev_threshold::Float64
+    current_time::Float64
+    current_threshold::Float64  
+    cumulative_rate::Float64
+    jumps::JumpSet
+    function GillespieIntegCallbackEventCache(jumps::JumpSet)
+        initial_threshold = -log(rand())
+        new(0.0, initial_threshold, 0.0, initial_threshold, 0.0, jumps)
+    end
+end
+
+# Condition function using 4-point Gaussian quadrature to determine event times
+struct GillespieIntegCallbackCondition <: Function
+    cache::GillespieIntegCallbackEventCache
+end
+
+function (cond::GillespieIntegCallbackCondition)(u, t, integrator)
+    cache = cond.cache
```

Comment:
Why not just define this directly on the `GillespieIntegCallbackEventCache`? i.e.
```julia
function (cond::GillespieIntegCallbackEventCache)(u, t, integrator)
...
end
function (aff:GillespieIntegCallbackEventCache)(integrator)
...
end
```
Why do you need to introduce new types to wrap these functions?